File: 백준/Gold/16234. 인구 이동/인구 이동.py
from collections import deque
import sys
input = sys.stdin.readline

N, L, R = map(int,input().split())
arr = [list(map(int,input().split())) for _ in range(N)]
dx = [-1,1,0,0]
dy = [0,0,-1,1]
queue = deque()
def show():
    for i in arr:
        print(' '.join(map(str,i)))

# show()
cnt = 0
while 1:
    
    visit = [[0 for _ in range(N)] for __ in range(N)]
    nomove = 0
    for i in range(N):
        for j in range(N):
            if visit[i][j] == 1:
                continue
            queue.append((i,j))
            group = set()
            group.add((i,j))
            groupsum = arr[i][j]
            visit[i][j] = 1
            while queue:
                x, y = queue.popleft()

                for k in range(4):
                    nx = dx[k] + x
                    ny = dy[k] + y
                    if nx < 0 or nx >= N or ny < 0 or ny >= N:
                        continue
                    if visit[nx][ny] == 1:
                        continue
                    if abs(arr[x][y] - arr[nx][ny]) >= L and abs(arr[x][y] - arr[nx][ny]) <= R:
                        group.add((nx,ny))
                        groupsum += arr[nx][ny]
                        visit[nx][ny] = 1
                        queue.append((nx,ny))
            if len(group) == 1:
                nomove += 1
                continue
            gn = groupsum // len(group)
            for g in group:
                x, y = g[0],g[1]
                arr[x][y] = gn
            # show()
    if nomove == N*N:
        print(cnt)
        break
    cnt += 1

File: 백준/Gold/16234. 인구 이동/인구 이동.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

byte[] buffer = new byte[1 << 20];
int bufferLength = 0, bufferIndex = 0;

byte ReadByte()
{
    if (bufferIndex == bufferLength)
    {
        bufferLength = Console.OpenStandardInput().Read(buffer, 0, 1 << 20);
        if (bufferLength == 0) return 255;
        bufferIndex = 0;
    }
    return buffer[bufferIndex++];
}
int ReadInt()
{
    int result = 0, sign = 1;
    byte b = ReadByte();
    while (b < '0' || b > '9')
    {
        if (b == '-') sign = -1;
        b = ReadByte();
    }
    while (b >= '0' && b <= '9')
    {
        result = result * 10 + (b - '0');
        b = ReadByte();
    }
    return result * sign;
}
char ReadChar()
{
    byte b = ReadByte();
    while (b <= ' ')
    {
        b = ReadByte();
    }
    return (char)b;
}

int N = ReadInt();
int L = ReadInt();
int R = ReadInt();
int[,] map = new int[N,N];
int[,] visit = new int[N,N];
int res = 0;
int cnt = 0;
int[] dx = new int[4] { -1, 1, 0, 0 };
int[] dy = new int[4] { 0, 0, -1, 1 };

for (int i = 0; i < N; i++)
{
    for (int j = 0; j < N; j++)
    {
        map[i, j] = ReadInt();
    }
}
Queue<Tuple<int, int>> queue = new Queue<Tuple<int, int>>();
Queue<Tuple<int, int>> group = new Queue<Tuple<int, int>>();
while (true)
{
    bool isMove = false;
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            if (visit[i, j] == 1) continue; 
            visit[i, j] = 1;
            queue.Enqueue(new Tuple<int, int>(i, j));
            group.Enqueue(new Tuple<int, int>(i, j));
            cnt += map[i, j];
            if (Move()) isMove = true;
        }
    }
    
    if (!isMove) break;
    
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            visit[i, j] = 0;
        }
    }
    res++;
}

Console.WriteLine(res);

bool Move()
{
    int x, y;
    
    while (queue.Count > 0)
    {
        (x, y) = queue.Dequeue();
        for (int i = 0; i < 4; i++)
        {
            int nx = x + dx[i];
            int ny = y + dy[i];
            if (nx < 0 || nx >= N || ny < 0 || ny >= N) continue;
            if (visit[nx, ny] == 1) continue;
            int abs = Math.Abs(map[nx, ny] - map[x, y]);
            if (abs < L || abs > R) continue;
            visit[nx, ny] = 1;
            queue.Enqueue(new Tuple<int, int>(nx, ny));
            group.Enqueue(new Tuple<int, int>(nx, ny));
            cnt += map[nx, ny];
        }
    }
    int gc = group.Count;

    if (gc == 1)
    {
        group.Clear();
        cnt = 0;
        return false;
    }
    
    while (group.Count > 0)
    {
        (x, y) = group.Dequeue();
        map[x, y] = cnt/gc;
    }

    cnt = 0;
    // PrintMap();
    return true;
}

void PrintMap()
{
    var sb = new StringBuilder();
    for (int i = 0; i < N; i++)
    {
        for (int j = 0; j < N; j++)
        {
            sb.Append(map[i, j] + " ");
        }
        sb.AppendLine();
    }
    Console.WriteLine(sb);
}

